### ISSUE TITLE
Ubuntu CMake: Cyclic dependencies are allowed only among static libraries.

### ISSUE THREAD

**Issue by Phil1988** [hyperion-project/hyperion.ng#511]
I'm trying to install hyperion.ng on my Odroid C2 with Ubuntu 16.04 64bit.

on CMake I get an error:

```
cmake -DCMAKE_BUILD_TYPE=Release ..
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Deprecation Warning at CMakeLists.txt:11 (CMAKE_POLICY):
  The OLD behavior for policy CMP0026 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- PLATFORM is not defined, evaluated platform: amlogic64
-- PLATFORM: amlogic64
-- ENABLE_AMLOGIC = ON
-- ENABLE_DISPMANX = OFF
-- ENABLE_FB = ON
-- ENABLE_OSX = OFF
-- ENABLE_SPIDEV   = ON
-- ENABLE_TINKERFORGE = ON
-- ENABLE_V4L2 = ON
-- ENABLE_WS281XPWM = OFF
-- ENABLE_USB_HID = ON
-- ENABLE_X11 = OFF
-- ENABLE_TESTS = OFF
-- ENABLE_PROFILER = OFF
-- validate json files
   checked files: 71 success: 71 errors: 0
-- validate json effect files
   checked effect files: 35 success: 35 errors: 0
-- validate json file
-- Current Version: master (GitHub-0049e23/e438bc6-1512140297)
-- Performing Test COMPILER_SUPPORTS_CXX11
-- Performing Test COMPILER_SUPPORTS_CXX11 - Success
-- Performing Test COMPILER_SUPPORTS_CXX0X
-- Performing Test COMPILER_SUPPORTS_CXX0X - Success
-- Found Qt Version: 5.5.1
-- Found libusb-1.0:
--  - Includes: /usr/include/libusb-1.0
--  - Libraries: /usr/lib/aarch64-linux-gnu/libusb-1.0.so
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found ZLIB: /usr/lib/aarch64-linux-gnu/libz.so (found version "1.2.8")
-- Looking for dlfcn.h
-- Looking for dlfcn.h - found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Looking for memory.h
-- Looking for memory.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stdlib.h
-- Looking for stdlib.h - found
-- Looking for strings.h
-- Looking for strings.h - found
-- Looking for string.h
-- Looking for string.h - found
-- Looking for sys/stat.h
-- Looking for sys/stat.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Protobuf: doing try-compiles for hash map/set headers
-- Protobuf: doing try-compiles for pthread test
-- Looking for stddef.h
-- Looking for stddef.h - found
CMake Deprecation Warning at dependencies/external/protobuf/src/CMakeLists.txt:3                                                                                                                                                             29 (cmake_policy):
  The OLD behavior for policy CMP0026 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- Using protobuf compiler: /root/hyperion/build/bin/protoc
-- Found PythonLibs: /usr/lib/aarch64-linux-gnu/libpython3.5m.so (found suitable                                                                                                                                                              version "3.5.2", minimum required is "3.4")
CMake Warning at doc/CMakeLists.txt:31 (message):
  Doxygen not found, unable to generate documenation!


-- Configuring done
CMake Error: The inter-target dependency graph contains the following strongly c                                                                                                                                                             onnected component (cycle):
  "hyperion" of type STATIC_LIBRARY
    depends on "blackborder" (weak)
    depends on "hyperion-utils" (weak)
    depends on "leddevice" (weak)
    depends on "bonjour" (weak)
    depends on "effectengine" (weak)
    depends on "hyperion_autogen" (strong)
  "hyperion_autogen" of type UTILITY
    depends on "blackborder" (strong)
    depends on "bonjour" (strong)
    depends on "effectengine" (strong)
    depends on "hyperion-utils" (strong)
    depends on "leddevice" (strong)
  "blackborder" of type STATIC_LIBRARY
    depends on "hyperion-utils" (weak)
    depends on "hyperion" (weak)
    depends on "leddevice" (weak)
    depends on "bonjour" (weak)
    depends on "effectengine" (weak)
    depends on "blackborder_autogen" (strong)
  "blackborder_autogen" of type UTILITY
    depends on "hyperion-utils" (strong)
  "bonjour" of type STATIC_LIBRARY
    depends on "hyperion" (weak)
    depends on "blackborder" (weak)
    depends on "hyperion-utils" (weak)
    depends on "leddevice" (weak)
    depends on "effectengine" (weak)
    depends on "bonjour_autogen" (strong)
  "bonjour_autogen" of type UTILITY
    depends on "hyperion" (strong)
    depends on "hyperion-utils" (strong)
  "leddevice" of type STATIC_LIBRARY
    depends on "hyperion-utils" (weak)
    depends on "hyperion" (weak)
    depends on "blackborder" (weak)
    depends on "bonjour" (weak)
    depends on "effectengine" (weak)
    depends on "leddevice_autogen" (strong)
  "leddevice_autogen" of type UTILITY
    depends on "hyperion-utils" (strong)
  "hyperion-utils" of type STATIC_LIBRARY
    depends on "hyperion" (weak)
    depends on "blackborder" (weak)
    depends on "leddevice" (weak)
    depends on "bonjour" (weak)
    depends on "effectengine" (weak)
    depends on "hyperion-utils_autogen" (strong)
  "hyperion-utils_autogen" of type UTILITY
    depends on "hyperion" (strong)
  "effectengine" of type STATIC_LIBRARY
    depends on "hyperion" (weak)
    depends on "blackborder" (weak)
    depends on "hyperion-utils" (weak)
    depends on "leddevice" (weak)
    depends on "bonjour" (weak)
    depends on "effectengine_autogen" (strong)
  "effectengine_autogen" of type UTILITY
    depends on "hyperion" (strong)
At least one of these targets is not a STATIC_LIBRARY.  Cyclic dependencies are                                                                                                                                                              allowed only among static libraries.
```

I found the same issue on arch but I have no idea how to solve this for my on Ubuntu:
https://github.com/hyperion-project/hyperion.ng/issues/482

I hope you guys can help me.

Thanks a lot!

**Comment by Phil1988** [comment]
As I found this: https://mariadb.atlassian.net/browse/MXS-128
it seems that CMake 3.9> will cause this error if the code is not set up correctly for it.

So in unstalled cmake and installed an older version (3.5.1).
Now I can run
 `cmake -DCMAKE_BUILD_TYPE=Release ..`
but as I try to to install it with `make -j $(nproc)`
I get another error at 90%:

`[ 90%] Building CXX object libsrc/webconfig/CMakeFiles/webconfig.dir/CMakeFiles/                                   webconfig.dir/qrc_WebConfig.cpp.o
Segmentation fault
libsrc/commandline/CMakeFiles/commandline.dir/build.make:254: recipe for target                                    'libsrc/commandline/CMakeFiles/commandline.dir/SwitchOption.cpp.o' failed
make[2]: *** [libsrc/commandline/CMakeFiles/commandline.dir/SwitchOption.cpp.o]                                    Error 139
CMakeFiles/Makefile2:788: recipe for target 'libsrc/commandline/CMakeFiles/comma                                   ndline.dir/all' failed
make[1]: *** [libsrc/commandline/CMakeFiles/commandline.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
Segmentation fault
libsrc/webconfig/CMakeFiles/webconfig.dir/build.make:380: recipe for target 'lib                                   /libwebconfig.a' failed
make[2]: *** [lib/libwebconfig.a] Error 139
CMakeFiles/Makefile2:1989: recipe for target 'libsrc/webconfig/CMakeFiles/webcon                                   fig.dir/all' failed
make[1]: *** [libsrc/webconfig/CMakeFiles/webconfig.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2`

**Comment by Phil1988** [comment]
I tried to install hyperion again by deleting the folder and install it "from scratch"

and it fails again at installing the v4l2-grabber

`
[ 87%] Built target v4l2-grabber
[ 87%] Building CXX object libsrc/webconfig/CMakeFiles/webconfig.dir/QtHttpReply.cpp.o
[ 88%] Building CXX object libsrc/webconfig/CMakeFiles/webconfig.dir/QtHttpRequest.cpp.o
[ 89%] Building CXX object libsrc/commandline/CMakeFiles/commandline.dir/ValidatorOption.cpp.o
[ 89%] Building CXX object libsrc/webconfig/CMakeFiles/webconfig.dir/StaticFileServing.cpp.o
[ 89%] Building CXX object libsrc/webconfig/CMakeFiles/webconfig.dir/webconfig_automoc.cpp.o
[ 90%] Building CXX object libsrc/webconfig/CMakeFiles/webconfig.dir/CMakeFiles/webconfig.dir/qrc_WebConfig.cpp.o
Segmentation fault
`

**Comment by penfold42** [comment]
You’re probably running out of ram during the build.

When it fails, try simply rerunning the make command.

If that doesn’t work try with make -j 1

**Comment by Phil1988** [comment]
Thanks @penfold42 for your help.

I did try to rerun the make but it always aborted at the exact same point... 
so it might be a ram issue but always at the same point is a hint that its not only ram...

I did serveral attempts trying to install it, mixing different tutorials and howTo's to get it installed.

Sadly I didnt get your answer in time so I havent tried the "make -j 1" command.


The current state is, that I have installed hyperion (not NG) on my Odroid C2 running Armbian_5.49_Odroidc2_Debian_stretch_next_4.17.3.

The problem that appears now is that the paths might be wrong.
`sudo /usr/local/bin/hyperiond restart`
gives me: ERROR: No valid config found
also there is a config in /etc/hyperion/

If I start hyperion by the command: 
`hyperiond /etc/hyperion/hyperion.config.json`
then the LEDs starts flickering. (no power issue.. there is still another problem)

If I use the android app, I can connect via the json server and set the color by the colorwheel.
Its working correct (no flickering) as long as I keep changing the color.
As son as I stop changing the color it goes back to the flickering.
Also there are no effects loaded to choose from.

My guess is, that the installation of hyperion has set some wrong paths... 
so the config and the effects cant be loaded automatically.
Probably also the flickering is caused by the same wrong paths.

A more detailed description of my problem is here:
https://hyperion-project.org/threads/odroid-c2-with-hyperion.3073/

I hope you or someone else can help me out :)

Thanks a lot!
